types/src/proof/accumulator/mock.rs

```diff
@@ -0,0 +1,174 @@
+// Copyright (c) The Diem Core Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::{
+    proof::{
+        position::{FrozenSubtreeSiblingIterator, Position},
+        AccumulatorConsistencyProof, MerkleTreeInternalNode, TransactionAccumulatorSummary,
+    },
+    transaction::Version,
+};
+use diem_crypto::hash::{
+    CryptoHash, HashValue, TransactionAccumulatorHasher, ACCUMULATOR_PLACEHOLDER_HASH,
+};
+use std::collections::BTreeMap;
+
+// TODO(philiphayes): another alternative might be to extract storage/accumulator
+// crate into diem-types.
+/// An immutable transaction accumulator (not a summary, since it stores _all_
+/// frozen internal nodes and leaf nodes) that allows for easily computing root
+/// hashes, consistency proofs, etc... at any versions in the range [0, num_leaves).
+///
+/// Note: intended for test-only code; panics whenever something goes wrong.
+#[derive(Clone, Debug)]
+pub struct MockTransactionAccumulator {
+    nodes: BTreeMap<Position, HashValue>,
```

Comment:
`HashMap` should suffice? And `Ord` and `PartialOrd` should not be needed.

----

Or if we really intend to utilize `Position::post_order_idx()`, a vector should do.

----
But I'd prefer not to use Position at all here, trying to be as different with production code as possible, and as straightforward as possible (examples: `NaiveSmt`, `NaiveSubtree` in the codebase.). Possible directions I could think of:
1. a real binary tree. `let level = leaves.map(create_node).collect(); while level.len() > 1 {level.chunks(2).map(create_internal_node).collect()}`
2. a vector of leaves and reply on recursive algorithms, probably plus a cache for performance (`HashMap<(HashValue, HashValue), HashValue)>`, for internal node hashing)
I guess it's easier saying than doing. So, your call. :P

Reply:
done! if you check out the MockTransactionAccumulator fixup! commit for the pure recursive implementation

Reply:
That is elegant, overwhelmingly. Thanks ❤️.